File: dart-to-javascript-interop-node/websockets/websocket_impl.js
const IsoWebSocket = require('isomorphic-ws');

class WebSocketServer {
  constructor(port, onConnection) {
    this.binding = new IsoWebSocket.Server({ port: port });
    this.binding.on('connection', function (socket, request) {
      onConnection(
        new WebSocket(socket),
        {
          'address': request.connection.remoteAddress,
          'port': request.connection.remotePort,
        }
      );
    });
  }

  close() {
    this.binding.close();
  }
}

class WebSocketClient {
  constructor(host, port, onConnected) {
    this.socket = new WebSocket(`ws://${host}:${port}`);
    this.socket.on('open', onConnected(new WebSocket(this.socket)));
  }

  close() {
    this.socket.close();
  }
}

class WebSocket {
  constructor(socket) {
    this.socket = socket;
  }

  send(data) {
    this.socket.send(data);
  }

  listen(onMessage) {
    this.socket.on('message', function message(data) {
      onMessage(data);
    });
  }

  close() {
    this.socket.close();
  }
}

module.exports = {
  WebSocketServer,
  WebSocketClient,
  WebSocket
};

File: dart-to-javascript-interop-node/websockets/websocket_impl.dart
@JS()
library websocket_js_impl;

import 'comms.dart';
import 'package:js/js.dart';

@JS()
class WebSocketServer implements CommsServer {
  external factory WebSocketServer(
    int port,
    void onConnection(WebSocket socket, ClientInfo client),
  );
  external void close();
}

@JS()
class WebSocketClient implements CommsClient {
  external factory WebSocketClient(
    String host,
    int port,
    void onConnected(WebSocket socket),
  );
  external void close();
}

@JS()
class WebSocket implements Socket {
  external factory WebSocket();
  external void send(data);
  external void listen(void onMessage(data));
  external void close();
}

@JS()
@anonymous
class WebSocketClientInfo implements ClientInfo {
  external String get address;
  external int get port;
  external factory WebSocketClientInfo({
    String address,
    int port,
  });
}
